### PR TITLE
Implement CreateContainer

### PR DESCRIPTION
**Pull request by haiyanmeng**
Signed-off-by: Haiyan Meng hmeng@redhat.com

### CODE REVIEW

**Comment by haiyanmeng** [comment]
@mrunalp , @runcom , PTAL. 
This is just the first version, I still have serverl TODOs to finish.
The logic can be reviewed.

**Comment by haiyanmeng** [comment]
@mrunalp , PTAL.
I marked the points I am not sure with `TODO(hmeng)`. There are totally four of them.

File: server/runtime.go [hunk 1]
```diff
@@ -156,8 +157,224 @@ func (s *Server) ListPodSandbox(context.Context, *pb.ListPodSandboxRequest) (*pb
...
+		if src == "" {
+			return nil, fmt.Errorf("Mount.HostPath is empty")
+		}
+
+		options := "rw"
+		if mount.GetReadonly() {
+			options = "ro"
+		}
+
+		//TODO(hmeng): how to use this info? Do we need to handle relabel a FS with Selinux?
```
Comment (mrunalp) [inline server/runtime.go:250]: This should be handled in create/prepare rootfs where we chcon -R the container rootfs.

File: server/server.go [hunk 2]
```diff
@@ -31,25 +31,51 @@ func New(runtimePath, sandboxDir, containerDir string) (*Server, error) {
...
+func (s *Server) hasSandbox(name string) bool {
+	_, ok := s.state.sandboxes[name]
+	return ok
+}
+
+func (s *sandbox) addContainer(c *container) {
+	s.containers = append(s.containers, c)
+}
+
+type container struct {
```
Comment (mrunalp) [inline server/server.go]: We should move this to the oci package.

File: server/server.go [hunk 3]
```diff
@@ -31,25 +31,51 @@ func New(runtimePath, sandboxDir, containerDir string) (*Server, error) {
...
+	name    string
+	logPath string
+	labels  map[string]string
+	sandbox string
```
Comment (mrunalp) [inline server/server.go]: This could be a pointer back to the sandbox struct.

Comment (haiyanmeng) [inline server/server.go]: Okay. Will fix it.

File: server/runtime.go [hunk 4]
```diff
@@ -156,8 +157,224 @@ func (s *Server) ListPodSandbox(context.Context, *pb.ListPodSandboxRequest) (*pb
...
+	image := imageSpec.GetImage()
+	if image == "" {
+		return nil, fmt.Errorf("CreateContainerRequest.ContainerConfig.Image.Image is empty")
+	}
+	fmt.Println(image)
+
+	// creates a spec Generator with the default spec.
+	specgen := generate.New()
+
+	//TODO(hmeng): ContainerConfig.Command should be only used by docker damon and ignored here, right?
```
Comment (mrunalp) [inline server/runtime.go]: Yes, that is correct. We can ignore Command in our code.

File: server/server.go (hunk 2 above)
Comment (haiyanmeng) [inline server/server.go]: Okay. Will fix it.

File: server/runtime.go (hunk 1 above)
Comment (haiyanmeng) [inline server/runtime.go:250]: @mrunalp , according to `api.pb.go`, `Mount.SelinuxRelabel` is specific to each mount. 
Does it mean that we are supposed to run `chcon -R` on each mount whose `SelinuxRelabel` is true? 
It seems that we may need do this via prestart hooks, right?

File: server/server.go (hunk 3 above)
Comment (haiyanmeng) [inline server/server.go]: @mrunalp , changing `container.sandbox` from `string` to `*sandbox`, and moving `container` into the `oci` package would result in cyclic import:

package `server` already `import "github.com/mrunalp/ocid/oci"` in `server.server.go`;
if we move `container` into the `oci` package, package `oci` would need to `import "github.com/mrunalp/ocid/server"` to refer `sandbox` inside the `container` struct.

File: server/runtime.go (hunk 1 above)
Comment (mrunalp) [inline server/runtime.go:250]: We only use on MountLabel for a container in OCI runtime spec. I'll make a PR to get this fixed upstream.

File: server/server.go [hunk 5]
```diff
@@ -31,25 +31,50 @@ func New(runtimePath, sandboxDir, containerDir string) (*Server, error) {
...
+
+func (s *sandbox) addContainer(c *container) {
+	s.containers = append(s.containers, c)
+}
+
+type container struct {
+	name    string
+	logPath string
+	labels  map[string]string
+	sandbox *sandbox
```
Comment (mrunalp) [inline server/server.go]: I think we can still move it to oci with a string for sandbox if we need a back reference.

Comment (haiyanmeng) [inline server/server.go]: Sounds good.

File: server/runtime.go (hunk 1 above)
Comment (haiyanmeng) [inline server/runtime.go:250]: > We only use on MountLabel for a container in OCI runtime spec.

Yes. 

> I'll make a PR to get this fixed upstream.

Great.

**Comment by haiyanmeng** [comment]
> I think we can still move it to oci with a string for sandbox if we need a back reference.

@mrunalp , PTAL. Fixed it.

File: oci/oci.go [hunk 6]
```diff
@@ -58,3 +59,49 @@ func getOCIVersion(name string, args ...string) (string, error) {
...
+// CreateContainer creates a container.
+func (r *Runtime) CreateContainer(bundle, name string) error {
+	if _, err := utils.ExecCmd(r.path, "--bundle", bundle, name); err != nil {
+		return fmt.Errorf("`%s --bundle %s %s` failed: %v", r.path, bundle, name, err)
+	}
+	return nil
+}
+
+// Container respresents a runtime container.
+type Container struct {
```
Comment (mrunalp) [inline oci/oci.go:69]: add bundlePath here

File: oci/oci.go [hunk 7]
```diff
@@ -58,3 +59,49 @@ func getOCIVersion(name string, args ...string) (string, error) {
 	v := firstLine[strings.LastIndex(firstLine, " ")+1:]
 	return v, nil
 }
+
+// CreateContainer creates a container.
+func (r *Runtime) CreateContainer(bundle, name string) error {
```
Comment (mrunalp) [inline oci/oci.go]: This could take a container object instead of individual fields.

File: oci/oci.go (hunk 6 above)
Comment (haiyanmeng) [inline oci/oci.go:69]: Fixed it.

File: oci/oci.go (hunk 7 above)
Comment (haiyanmeng) [inline oci/oci.go]: Fixed it.

**Comment by haiyanmeng** [comment]
@mrunalp , I tried `redis` and `hello-world`, and unfortunately the same problem still happened.
I can run `sudo $(which runc) create --bundle /var/lib/ocid/containers/container1/ container1` successfully with the `config.json` created by `ocic createcontainer`.
However, wrapping the runc command inside `exec.Command` somehow failed `runc create`: `container1` got created, but `ocic createcontainer` just hanged up forever.

**Comment by haiyanmeng** [comment]
@mrunalp , I put the test `container_config.json` file under `ocid/testdata` and also put a `README.md` there which includes the test command.
If you have time, you may want to check the `container_config.json` file there.

**Comment by haiyanmeng** [comment]
@mrunalp , I think I found the problem.
The behavior should be caused by the `fork/exec` model of golang.
Because If I set the `-D` flag of strace, I can strace busybox without tty successfully.

```
[hmeng@localhost c4]$ sudo strace -f -o log -D $(which runc) create 4
```

However, if I do not use the `-D` flag of strace, stracing busybox without tty failed:

```
[hmeng@localhost c4]$ sudo strace -f -o log1 $(which runc) create 5
```

**Comment by mrunalp** [comment]
@hmeng-19 Are we setting cmd.Stdin to os.Stdin cmd.Stdout to os.Stdout and cmd.Stderr to os.Stderr?

**Comment by haiyanmeng** [comment]
@mrunalp , we currently does not set `cmd.Stdin`, but we use `bytes.Buffer` to track `cmd.Stdout` and `cmd.Stderr`.

```
func ExecCmd(name string, args ...string) (string, error) {
    cmd := exec.Command(name, args...)
    var stdout bytes.Buffer
    var stderr bytes.Buffer
    cmd.Stdout = &stdout
    cmd.Stderr = &stderr

    err := cmd.Run()
    if err != nil {
        return "", fmt.Errorf("`%v %v` failed: %v (%v)", name, strings.Join(args, " "), stderr.String(), err)
    }

    return stdout.String(), nil
}
```

**Comment by haiyanmeng** [comment]
@mrunalp , fixed it. It seems setting cmd.Stdin|out|err to os.Stdin|out|err really solve the problem I have, even if I do not fully understand why it solved the problem.

**Comment by mrunalp** [comment]
@hmeng-19 cool, I'll test this out later today.

**Comment by haiyanmeng** [comment]
@mrunalp , sounds good. `ocid/testdata/README.md` contains some very basic test commands, which you may find useful for your test.
You are supposed to run the test commads in `ocid/testdata/README.md` under `ocid` directly.

**Comment by mrunalp** [comment]
@hmeng-19 Could you add the call to create container for pod sandbox as well in a different PR? That way we can add code to join its namespaces in this code path. We can use a specific image like the kubernetes pause image for the pod sandbox container.

**Comment by haiyanmeng** [comment]
@mrunalp , okay. I will do it after you are done testing and merging the PR.

**Comment by mrunalp** [comment]
LGTM. Good work!

**Pull request merged**